File: ScriptGraph/PinToGraph/PropertyScript/JS/Property.js
const Amaz = effect.Amaz;

const map = {};

class Property {
  constructor() {}

  setProperty(objects, property, value) {
    if (!Array.isArray(objects) || objects.length === 0) {
      return;
    }
    if (property === 'localEulerAngle') {
      map[objects[0].guid.toString()] = value;
    } else if (objects[0] instanceof Amaz.FaceStretchComponent && property === 'Intensity') {
      objects[0][property] = value / 100.0;
      return;
    }
    objects[0][property] = value;
  }

  getProperty(objects, property) {
    if (!Array.isArray(objects) || objects.length === 0) {
      return null;
    }
    if (property === 'localPosition') {
      const position = objects[0][property];
      return new Amaz.Vector3f(position.x, position.y, position.z);
    } else if (property === 'localEulerAngle') {
      if (map[objects[0].guid.toString()] !== undefined) {
        return map[objects[0].guid.toString()];
      }
      return objects[0][property];
    } else if (objects[0] instanceof Amaz.FaceStretchComponent && property === 'Intensity') {
      return objects[0][property] * 100;
    }
    return objects[0][property];
  }
}

exports.Property = Property;

File: ScriptGraph/PinToGraph/PropertyScript/Lua/Property.lua
local Property = Property or {}
Property.__index = Property

local map = {}

function Property.new()
  local self = setmetatable({}, Property)
  return self
end

function Property:setProperty(objects, property, value)
  if #objects == 0 then
    return
  end
  if property == "localEulerAngle" then
    map[tostring(objects[1].guid)] = value
  end
  objects[1][property] = value
end

function Property:getProperty(objects, property)
  if #objects == 0 then
    return nil
  end
  if property == "localPosition" then
    local position = objects[1][property]
    return Amaz.Vector3f(position.x, position.y, position.z)
  elseif property == "localEulerAngle" then
    if map[tostring(objects[1].guid)] ~= nil then
      return map[tostring(objects[1].guid)]
    end
    return objects[1].localEulerAngle
  end
  return objects[1][property]
end

return Property